src/model/wallet.js

```diff
@@ -831,3 +842,127 @@ export const loadSeedPhraseAndMigrateIfNeeded = async id => {
     return null;
   }
 };
+
+async function extractSecretsForWallet(wallet) {
+  const allKeys = await keychain.loadAllKeys();
+  if (!allKeys) throw new Error("Couldn't read secrets from keychain");
+  const secrets = {};
+
+  const allowedPkeysKeys = wallet.addresses.map(
+    account => `${account.address}_${privateKeyKey}`
+  );
+
+  allKeys.forEach(item => {
+    // Ignore allWalletsKey
+    if (item.username === allWalletsKey) {
+      return;
+    }
+
+    // Ignore selected wallet
+    if (item.username === selectedWalletKey) {
+      return;
+    }
+
+    // Ignore icloud backup password
+    if (item.username === 'rainbowBackup') {
+      return;
+    }
+
+    // Ignore another wallets seeds
+    if (
+      item.username.indexOf(`_${seedPhraseKey}`) !== -1 &&
+      item.username !== `${wallet.id}_${seedPhraseKey}`
+    ) {
+      return;
+    }
+
+    // Ignore other wallets PKeys
+    if (
+      item.username.indexOf(privateKeyKey) !== -1 &&
+      allowedPkeysKeys.indexOf(item.username) === -1
+    ) {
+      return;
+    }
+
+    secrets[item.username] = item.password;
+  });
+  return secrets;
+}
+
+export async function backupWalletToCloud(password, wallet) {
+  logger.log('backupWalletToCloud', password, wallet);
+  const now = Date.now();
+
+  const secrets = await extractSecretsForWallet(wallet);
+  logger.log('backupWalletToCloud:: got secrets', secrets);
+  const data = {
+    createdAt: now,
+    secrets,
+  };
+  logger.log('backupWalletToCloud:: about to encrypt and save data', data);
+  return encryptAndSaveDataToCloud(data, password, `backup_${now}.json`);
+}
+
+export async function addWalletToCloudBackup(password, wallet, filename) {
+  logger.log('addWalletToCloudBackup', password, wallet, filename);
+
+  const backup = await getDataFromCloud(password, filename);
```

Comment:
getDataFromCloud could return null when it has an internal error, so should do some checks on backup

Reply:
There's already UI error handling everywhere where it's being called:

- https://github.com/rainbow-me/rainbow/pull/771/files#diff-8b5c7a88c66039ee7963595dac3a9ef0R115
- https://github.com/rainbow-me/rainbow/pull/771/files#diff-0003d63a111d13248e261d99c808df36R224
- https://github.com/rainbow-me/rainbow/pull/771/files#diff-eddf778cb05c614e22ad92e3b1014d78R302
- https://github.com/rainbow-me/rainbow/pull/771/files#diff-af5c0ea58fa152bce42f3e49dd35fd3dR151

is that what you mean?

Reply:
Oh I mean the const `backup` in line 867 can be null, and we are continuing on with the logic (later setting backup.updatedAt in line 873) - we should check if backup is not nil before continuing (either throw an error or return early if it is nil value)

Reply:
Made a separate comment, so resolving this one